File: src/components/world/chunk/TerrainGenerator.js
import { BLOCK } from './TextureAtlas';

// Simple Simplex Noise 2D Implementation
class SimplexNoise2D {
  constructor(seed = 4242) {
    this.p = new Uint8Array(256);
    this.perm = new Uint8Array(512);
    this.permMod12 = new Uint8Array(512);

    for (let i = 0; i < 256; i++) this.p[i] = i;
    for (let i = 255; i > 0; i--) {
      const j = Math.floor(Math.abs(Math.sin(seed + i) * 10000)) % (i + 1);
      const tmp = this.p[i];
      this.p[i] = this.p[j];
      this.p[j] = tmp;
    }
    for (let i = 0; i < 512; i++) {
      this.perm[i] = this.p[i & 255];
      this.permMod12[i] = this.perm[i] % 12;
    }
    this.G2 = (3.0 - Math.sqrt(3.0)) / 6.0;
    this.F2 = 0.5 * (Math.sqrt(3.0) - 1.0);
    this.grad3 = [
      [1, 1, 0], [-1, 1, 0], [1, -1, 0], [-1, -1, 0],
      [1, 0, 1], [-1, 0, 1], [1, 0, -1], [-1, 0, -1],
      [0, 1, 1], [0, -1, 1], [0, 1, -1], [0, -1, -1]
    ];
  }

  noise(xin, yin) {
    const s = (xin + yin) * this.F2;
    const i = Math.floor(xin + s);
    const j = Math.floor(yin + s);
    const t = (i + j) * this.G2;
    const X0 = i - t;
    const Y0 = j - t;
    const x0 = xin - X0;
    const y0 = yin - Y0;

    let i1, j1;
    if (x0 > y0) { i1 = 1; j1 = 0; } else { i1 = 0; j1 = 1; }

    const x1 = x0 - i1 + this.G2;
    const y1 = y0 - j1 + this.G2;
    const x2 = x0 - 1.0 + 2.0 * this.G2;
    const y2 = y0 - 1.0 + 2.0 * this.G2;

    const ii = i & 255;
    const jj = j & 255;
    const gi0 = this.permMod12[ii + this.perm[jj]];
    const gi1 = this.permMod12[ii + i1 + this.perm[jj + j1]];
    const gi2 = this.permMod12[ii + 1 + this.perm[jj + 1]];

    let n0 = 0, n1 = 0, n2 = 0;
    let t0 = 0.5 - x0 * x0 - y0 * y0;
    if (t0 > 0) { t0 *= t0; n0 = t0 * t0 * (this.grad3[gi0][0] * x0 + this.grad3[gi0][1] * y0); }
    let t1 = 0.5 - x1 * x1 - y1 * y1;
    if (t1 > 0) { t1 *= t1; n1 = t1 * t1 * (this.grad3[gi1][0] * x1 + this.grad3[gi1][1] * y1); }
    let t2 = 0.5 - x2 * x2 - y2 * y2;
    if (t2 > 0) { t2 *= t2; n2 = t2 * t2 * (this.grad3[gi2][0] * x2 + this.grad3[gi2][1] * y2); }

    return 70.0 * (n0 + n1 + n2);
  }
}

export class TerrainGenerator {
  constructor(seed = 424242) {
    this.noiseBase = new SimplexNoise2D(seed);
    this.noiseDetail = new SimplexNoise2D(seed + 1337);
    this.noiseMountain = new SimplexNoise2D(seed + 9999);
    this.seaLevel = 4; // Water level at Y=4

    this.pois = [
      { id: 'spawn_lectern', name: "Adventurer's Lectern", subtitle: "River Clearing", icon: "📜", x: 6, z: -10, excerpt: "A hand-carved oak pedestal holding an open journal." },
      { id: 'crystal_lake', name: "Crystal Lake Study", subtitle: "Mirror Water's Edge", icon: "🌊", x: 14, z: 28, excerpt: "A tranquil research post set up beside clear spring waters." },
      { id: 'shrine', name: "Forgotten Stone Shrine", subtitle: "Eastern Overlook", icon: "🏛️", x: 22, z: 16, excerpt: "An ancient stone altar flanked by moss-weathered pillars." },
      { id: 'wanderer_cache', name: "Wanderer's Cache", subtitle: "Dense Birch Border", icon: "📦", x: -18, z: -24, excerpt: "A sturdy chest resting beside a cluster of oak logs." },
      { id: 'mineshaft_entrance', name: "Abandoned Mine Entrance", subtitle: "Deep Ravine Edge", icon: "⛏️", x: 32, z: -28, excerpt: "Weathered oak beam scaffolding leading into a dark mineral-rich tunnel." },
      { id: 'deep_ridge', name: "Deep Ridge Mineral Pocket", subtitle: "Southern Granite Crags", icon: "💎", x: -26, z: 32, excerpt: "Exposed veins of blue diamond ore glittering under the mountain sun." }
    ];
  }

  computeHeight(wx, wz) {
    // 1. Winding River Valley
    const riverPath = Math.sin(wz * 0.045) * 14 + Math.cos(wz * 0.02) * 5;
    const distToRiver = Math.abs(wx - riverPath);

    if (distToRiver < 5.0) {
      return 2; // River bed (covered with water up to Y=4)
    }

    if (distToRiver < 7.5) {
      const t = (distToRiver - 5.0) / 2.5;
      return Math.round(2 + t * 2);
    }

    // 2. Rolling Plains & Gentle Hills
    const n1 = this.noiseBase.noise(wx * 0.025, wz * 0.025) * 3.5;
    const n2 = this.noiseDetail.noise(wx * 0.08, wz * 0.08) * 1.5;
    const m = Math.max(0, this.noiseMountain.noise(wx * 0.012, wz * 0.012)) * 5.0;

    let h = Math.round(5.0 + n1 + n2 + m);
    return Math.max(3, Math.min(22, h));
  }

  generateChunkData(cx, cz) {
    const size = 16;
    const height = 32;
    const blocks = new Uint8Array(size * size * height);

    const setBlock = (x, y, z, blockType) => {
      if (x >= 0 && x < size && z >= 0 && z < size && y >= 0 && y < height) {
        blocks[(y * size + z) * size + x] = blockType;
      }
    };

    // 1. Terrain Strata Generation
    for (let lz = 0; lz < size; lz++) {
      for (let lx = 0; lx < size; lx++) {
        const wx = cx * size + lx;
        const wz = cz * size + lz;
        const surfaceY = this.computeHeight(wx, wz);

        // Bedrock
        setBlock(lx, 0, lz, BLOCK.BEDROCK);

        // Underground Strata
        for (let y = 1; y < surfaceY; y++) {
          if (y < surfaceY - 3) {
            const oreNoise = this.noiseDetail.noise(wx * 0.25, wz * 0.25 + y * 0.15);
            if (oreNoise > 0.62 && y <= 4) {
              setBlock(lx, y, lz, BLOCK.DIAMOND_ORE);
            } else {
              setBlock(lx, y, lz, BLOCK.STONE);
            }
          } else {
            setBlock(lx, y, lz, BLOCK.DIRT);
          }
        }

        // River Bed, Shoreline & Surface
        if (surfaceY <= 2) {
          // River Sand Bed
          setBlock(lx, surfaceY, lz, BLOCK.SAND);
          for (let wy = surfaceY + 1; wy <= this.seaLevel; wy++) {
            setBlock(lx, wy, lz, BLOCK.WATER);
          }
        } else if (surfaceY <= 3) {
          // River Shoreline Sand
          setBlock(lx, surfaceY, lz, BLOCK.SAND);
          if (surfaceY < this.seaLevel) {
            for (let wy = surfaceY + 1; wy <= this.seaLevel; wy++) {
              setBlock(lx, wy, lz, BLOCK.WATER);
            }
          }
        } else if (surfaceY > 11) {
          setBlock(lx, surfaceY, lz, BLOCK.STONE);
        } else {
          // Lush Green Grass Block
          setBlock(lx, surfaceY, lz, BLOCK.GRASS);

          // Subtle Wildflower & Grass Scatter (Vanilla 1.19 Rarity)
          const vegNoise = this.noiseDetail.noise(wx * 0.35, wz * 0.35);
          if (surfaceY + 1 < height) {
            if (vegNoise > 0.84) {
              setBlock(lx, surfaceY + 1, lz, BLOCK.POPPY);
            } else if (vegNoise > 0.76) {
              setBlock(lx, surfaceY + 1, lz, BLOCK.DANDELION);
            } else if (vegNoise > 0.58) {
              setBlock(lx, surfaceY + 1, lz, BLOCK.TALL_GRASS);
            }
          }
        }
      }
    }

    // 2. Dense Volumetric Oak & White Birch Forest Canopies
    const plantTree = (tx, tz, isBirch) => {
      const wx = cx * size + tx;
      const wz = cz * size + tz;
      const ty = this.computeHeight(wx, wz);

      if (ty >= 4 && ty <= 10) {
        const logType = isBirch ? BLOCK.BIRCH_LOG : BLOCK.OAK_LOG;
        const leafType = isBirch ? BLOCK.BIRCH_LEAVES : BLOCK.OAK_LEAVES;
        const trunkHeight = isBirch ? 5 : 4;

        setBlock(tx, ty + 1, tz, BLOCK.AIR);

        // Trunk
        for (let y = 1; y <= trunkHeight; y++) {
          setBlock(tx, ty + y, tz, logType);
        }

        // Volumetric 5x5 Lower Canopy (Full solid leaves, clipped outer corners)
        const leafBase = ty + trunkHeight - 1;
        for (let dx = -2; dx <= 2; dx++) {
          for (let dz = -2; dz <= 2; dz++) {
            if (Math.abs(dx) === 2 && Math.abs(dz) === 2) continue;
            setBlock(tx + dx, leafBase, tz + dz, leafType);
            setBlock(tx + dx, leafBase + 1, tz + dz, leafType);
          }
        }

        // Volumetric 3x3 Upper Canopy
        for (let dx = -1; dx <= 1; dx++) {
          for (let dz = -1; dz <= 1; dz++) {
            setBlock(tx + dx, leafBase + 2, tz + dz, leafType);
          }
        }

        // Top Rounded Leaf Crown (+ shape)
        setBlock(tx, leafBase + 3, tz, leafType);
        setBlock(tx + 1, leafBase + 3, tz, leafType);
        setBlock(tx - 1, leafBase + 3, tz, leafType);
        setBlock(tx, leafBase + 3, tz + 1, leafType);
        setBlock(tx, leafBase + 3, tz - 1, leafType);
      }
    };

    const seedHash = Math.abs((cx * 31 + cz) ^ 0x5DEECE66);
    if (seedHash % 2 === 0) {
      const t1x = (seedHash >> 2) % 8 + 4;
      const t1z = (seedHash >> 5) % 8 + 4;
      plantTree(t1x, t1z, seedHash % 4 === 0);

      const t2x = ((seedHash >> 7) % 6) + 2;
      const t2z = ((seedHash >> 9) % 6) + 8;
      plantTree(t2x, t2z, (seedHash >> 3) % 4 === 0);
    }

    return blocks;
  }
}

File: engine/src/com/mcjournal/TerrainGenerator.java
package com.mcjournal;

import java.util.*;

public class TerrainGenerator {
    public static final int SEA_LEVEL = 62; // Minecraft 1.17 Standard Sea Level
    public static final int BASE_GROUND_LEVEL = 64; // Minecraft 1.17 Plains Ground Level

    private final SimplexNoise baseNoise;
    private final SimplexNoise detailNoise;
    private final SimplexNoise mountainNoise;
    private final SimplexNoise caveNoise1;
    private final SimplexNoise caveNoise2;
    private final List<POI> pois = new ArrayList<>();

    public record POI(String id, String name, String subtitle, String icon, int x, int z, String excerpt) {}

    public TerrainGenerator(long seed) {
        this.baseNoise = new SimplexNoise(seed);
        this.detailNoise = new SimplexNoise(seed + 1337);
        this.mountainNoise = new SimplexNoise(seed + 9999);
        this.caveNoise1 = new SimplexNoise(seed + 4242);
        this.caveNoise2 = new SimplexNoise(seed + 8484);
        initPOIs();
    }

    private void initPOIs() {
        pois.add(new POI("spawn_lectern", "Adventurer's Lectern", "River Clearing", "📜", 6, -10, "A hand-carved oak pedestal holding an open journal."));
        pois.add(new POI("crystal_lake", "Crystal Lake Study", "Mirror Water's Edge", "🌊", 14, 28, "A tranquil research post set up beside clear spring waters."));
        pois.add(new POI("shrine", "Forgotten Stone Shrine", "Eastern Overlook", "🏛️", 22, 16, "An ancient stone altar flanked by moss-weathered pillars."));
        pois.add(new POI("wanderer_cache", "Wanderer's Cache", "Dense Birch Border", "📦", -18, -24, "A sturdy chest resting beside a cluster of oak logs."));
        pois.add(new POI("mineshaft_entrance", "Abandoned Mine Entrance", "Deep Ravine Edge", "⛏️", 32, -28, "Weathered oak beam scaffolding leading into a dark mineral-rich tunnel."));
        pois.add(new POI("deep_ridge", "Deep Ridge Mineral Pocket", "Southern Granite Crags", "💎", -26, 32, "Exposed veins of blue diamond ore glittering under the mountain sun."));
    }

    public List<POI> getPois() {
        return Collections.unmodifiableList(pois);
    }

    /**
     * Minecraft 1.17 Multi-Octave Continental & Mountain Heightmap.
     * Produces plains at Y=64-72, rolling hills at Y=73-85, rivers down to Y=52, and mountains up to Y=125.
     */
    public int computeHeight(int wx, int wz) {
        // Continental Base Noise (Plains & Gentle Rolling Hills)
        double continental = baseNoise.noise2D(wx * 0.012, wz * 0.012) * 8.0;
        double detail = detailNoise.noise2D(wx * 0.035, wz * 0.035) * 3.5;

        // Mountain Ridge Noise (Cliffs & Peaks)
        double ridge = Math.max(0, mountainNoise.noise2D(wx * 0.008, wz * 0.008));
        double mountain = Math.pow(ridge, 2.2) * 45.0;

        // River Channel Carving
        double riverCurve = Math.sin(wz * 0.025) * 22.0 + Math.cos(wz * 0.012) * 8.0;
        double distToRiver = Math.abs(wx - riverCurve);

        double baseHeight = BASE_GROUND_LEVEL + continental + detail + mountain;

        if (distToRiver < 7.0) {
            // Deep river bed (Y = 52 to 56)
            double depth = (1.0 - (distToRiver / 7.0));
            baseHeight -= depth * 16.0;
        } else if (distToRiver < 11.0) {
            // River bank slope
            double slope = (1.0 - ((distToRiver - 7.0) / 4.0));
            baseHeight -= slope * 6.0;
        }

        return Math.clamp((int) Math.round(baseHeight), 45, 230);
    }

    public Chunk generateChunk(int cx, int cz) {
        Chunk chunk = new Chunk(cx, cz);
        Random rng = new Random((long) cx * 341873128712L + (long) cz * 132897987541L);

        for (int lz = 0; lz < 16; lz++) {
            for (int lx = 0; lx < 16; lx++) {
                int wx = cx * 16 + lx;
                int wz = cz * 16 + lz;

                int surfaceY = computeHeight(wx, wz);

                // 1. Bedrock Floor (Y = 0 with random bedrock bumps up to Y = 3)
                chunk.setBlock(lx, 0, lz, Block.BEDROCK);
                for (int by = 1; by <= 3; by++) {
                    if (rng.nextFloat() < (1.0f - by * 0.28f)) {
                        chunk.setBlock(lx, by, lz, Block.BEDROCK);
                    }
                }

                // 2. Underground Mantle (Stone, Ores & 3D Caves)
                for (int y = 1; y < surfaceY; y++) {
                    if (chunk.getBlock(lx, y, lz) == Block.BEDROCK) continue;

                    if (y < surfaceY - 4) {
                        // 3D Minecraft 1.17 Cave & Tunnel Carving
                        if (y >= 5) {
                            double c1 = caveNoise1.noise3D(wx * 0.042, y * 0.062, wz * 0.042);
                            double c2 = caveNoise2.noise3D(wx * 0.042, y * 0.062, wz * 0.042);

                            // Noodle / Worm Tunnels & Spacious Caverns
                            double tunnelMetric = c1 * c1 + c2 * c2;
                            boolean isWormCave = (tunnelMetric < 0.016);
                            boolean isCavern = (c1 > 0.60 && Math.abs(c2) < 0.32);

                            if (isWormCave || isCavern) {
                                // Deep underground spring pools below Y = 12
                                if (y <= 11 && y >= 9 && !isWormCave) {
                                    chunk.setBlock(lx, y, lz, Block.WATER);
                                } else {
                                    chunk.setBlock(lx, y, lz, Block.AIR);
                                }
                                continue;
                            }
                        }

                        // Diamond Ore (Minecraft 1.17: Y = 1 to 16, peak density near bedrock)
                        if (y <= 16) {
                            double diamondNoise = detailNoise.noise2D(wx * 0.45, wz * 0.45 + y * 0.3);
                            if (diamondNoise > 0.72) {
                                chunk.setBlock(lx, y, lz, Block.DIAMOND_ORE);
                                continue;
                            }
                        }

                        // Cobblestone vein clusters in deep stone
                        double cobbleNoise = detailNoise.noise2D(wx * 0.15, wz * 0.15 + y * 0.1);
                        if (cobbleNoise > 0.75) {
                            chunk.setBlock(lx, y, lz, Block.COBBLESTONE);
                        } else {
                            chunk.setBlock(lx, y, lz, Block.STONE);
                        }
                    } else {
                        // Subsurface Dirt / Sand Layer (3-4 blocks below surface)
                        if (surfaceY <= SEA_LEVEL + 1) {
                            chunk.setBlock(lx, y, lz, Block.SAND);
                        } else {
                            chunk.setBlock(lx, y, lz, Block.DIRT);
                        }
                    }
                }

                // 3. Surface & Shoreline Layering
                if (surfaceY <= SEA_LEVEL) {
                    // Underwater River / Ocean Floor
                    chunk.setBlock(lx, surfaceY, lz, Block.SAND);

                    // Water column filling up to Sea Level (Y = 62)
                    for (int wy = surfaceY + 1; wy <= SEA_LEVEL; wy++) {
                        chunk.setBlock(lx, wy, lz, Block.WATER);
                    }
                } else if (surfaceY <= SEA_LEVEL + 2) {
                    // Beach / Shoreline
                    chunk.setBlock(lx, surfaceY, lz, Block.SAND);
                } else if (surfaceY > 105) {
                    // High Mountain Cliff / Rocky Peaks
                    chunk.setBlock(lx, surfaceY, lz, Block.STONE);
                } else {
                    // Lush Plains / Forest Surface (Grass Block)
                    chunk.setBlock(lx, surfaceY, lz, Block.GRASS);

                    // 4. Wildflower & Foliage Scatter
                    double vegNoise = detailNoise.noise2D(wx * 0.28, wz * 0.28);
                    if (surfaceY + 1 < Chunk.HEIGHT) {
                        if (vegNoise > 0.83) {
                            chunk.setBlock(lx, surfaceY + 1, lz, Block.POPPY);
                        } else if (vegNoise > 0.76) {
                            chunk.setBlock(lx, surfaceY + 1, lz, Block.DANDELION);
                        } else if (vegNoise > 0.52) {
                            chunk.setBlock(lx, surfaceY + 1, lz, Block.TALL_GRASS);
                        }
                    }
                }
            }
        }

        // 5. Forest Tree Generation (Oak & Birch on grass terrain)
        int seedHash = Math.abs((cx * 31 + cz) ^ 0x5DEECE66);
        if (seedHash % 3 <= 1) {
            int t1x = (seedHash >> 2) % 8 + 4;
            int t1z = (seedHash >> 5) % 8 + 4;
            plantTree(chunk, cx, cz, t1x, t1z, seedHash % 2 == 0);

            if (seedHash % 3 == 0) {
                int t2x = ((seedHash >> 7) % 6) + 2;
                int t2z = ((seedHash >> 9) % 6) + 8;
                plantTree(chunk, cx, cz, t2x, t2z, seedHash % 4 == 0);
            }
        }

        return chunk;
    }

    private void plantTree(Chunk chunk, int cx, int cz, int tx, int tz, boolean isBirch) {
        int wx = cx * 16 + tx;
        int wz = cz * 16 + tz;
        int ty = computeHeight(wx, wz);

        // Only plant on grass above sea level and below high mountain peaks
        if (ty >= SEA_LEVEL + 2 && ty <= 95 && chunk.getBlock(tx, ty, tz) == Block.GRASS) {
            byte logType = isBirch ? Block.BIRCH_LOG : Block.OAK_LOG;
            byte leafType = isBirch ? Block.BIRCH_LEAVES : Block.OAK_LEAVES;
            int trunkHeight = isBirch ? 5 : 4;

            // Clear any foliage above the ground
            chunk.setBlock(tx, ty + 1, tz, Block.AIR);

            // Wood Trunk
            for (int y = 1; y <= trunkHeight; y++) {
                if (ty + y < Chunk.HEIGHT) {
                    chunk.setBlock(tx, ty + y, tz, logType);
                }
            }

            // Canopy Leaves (3D Minecraft tree canopy)
            int leafBase = ty + trunkHeight - 1;
            for (int dx = -2; dx <= 2; dx++) {
                for (int dz = -2; dz <= 2; dz++) {
                    if (Math.abs(dx) == 2 && Math.abs(dz) == 2) continue;
                    if (tx + dx >= 0 && tx + dx < 16 && tz + dz >= 0 && tz + dz < 16) {
                        if (leafBase < Chunk.HEIGHT) chunk.setBlock(tx + dx, leafBase, tz + dz, leafType);
                        if (leafBase + 1 < Chunk.HEIGHT) chunk.setBlock(tx + dx, leafBase + 1, tz + dz, leafType);
                    }
                }
            }
            for (int dx = -1; dx <= 1; dx++) {
                for (int dz = -1; dz <= 1; dz++) {
                    if (tx + dx >= 0 && tx + dx < 16 && tz + dz >= 0 && tz + dz < 16) {
                        if (leafBase + 2 < Chunk.HEIGHT) chunk.setBlock(tx + dx, leafBase + 2, tz + dz, leafType);
                    }
                }
            }
            if (leafBase + 3 < Chunk.HEIGHT) {
                chunk.setBlock(tx, leafBase + 3, tz, leafType);
                if (tx + 1 < 16) chunk.setBlock(tx + 1, leafBase + 3, tz, leafType);
                if (tx - 1 >= 0) chunk.setBlock(tx - 1, leafBase + 3, tz, leafType);
                if (tz + 1 < 16) chunk.setBlock(tx, leafBase + 3, tz + 1, leafType);
                if (tz - 1 >= 0) chunk.setBlock(tx, leafBase + 3, tz - 1, leafType);
            }
        }
    }
}
